packages/styled-components/src/utils/flatten.js

```diff
@@ -55,23 +56,14 @@ export default function flatten(chunk: any, executionContext: ?Object, styleShee
 
   /* Either execute or defer the function */
   if (isFunction(chunk)) {
-    if (executionContext) {
-      let shouldThrow = false;
+    if (isStatelessFunction(chunk) && executionContext) {
+      const result = chunk(executionContext);
 
-      try {
-        // eslint-disable-next-line new-cap
-        if (isElement(new chunk(executionContext))) {
-          shouldThrow = true;
-        }
-      } catch (e) {
-        /* */
-      }
-
-      if (shouldThrow) {
+      if (isElement(result)) {
```

Comment:
This doesn't quite seem safe. We're calling a function that might be a component and are doing so to throw a proper warning if it is;

I don't think I even agree with the previous error to be honest. A return value of a component might not always be an element, so this doesn't seem safe.

It's also unsafe to simply call this because components might contain hooks now, so we'd still need the try-catch: https://runkit.com/kitten/5c7f0daf31404d0011b3a130

@probablyup @mxstbr Are we sure we really need this error in production? Seems a little overzealous. Maybe an error in development or even just a `console.error` would do the trick?

@alansouzati btw, can you change the changelog entry to say it's reducing memory usage when running component selector styles? It seems a little extreme to call it a memory leak, since I don't think this is leaking, it simply seems to be allocating a lot of memory that the GC is slow to clean up 😅 Still bad, but I'd hate to confuse people tbh

Reply:
sure I can re-word the change log. I will wait to do that until you decide if you want the check or not. Then I will update the PR in one go.

Reply:
The goal of this PR was to maintain the same features while fixing an important issue that is affecting performance.

Reply:
Is there an easy way to tell if something is a styled component or not ? the check could be changed to validate that instead.

Reply:
Yeah we have a util for it

Reply:
Cool does it make sense to use that instead? Unfortunately I don’t have a lot of the context for this.

Reply:
We check if it's a styled component at https://github.com/styled-components/styled-components/pull/2428/files#diff-ce0d3892db66efa4deef4063f44d3022R53 so not necessary

Reply:
I'm confused. can we remove the check or not?

Reply:
I'm just responding to your earlier question, interpolated styled components are handled earlier in `flatten` so you don't have to worry about them in the try/catch

Reply:
To @kitten's point, I think you're right and we can just do a dev-time warning & swallow the error in prod.

Reply:
is it ok to do that in a separate PR?

Reply:
I'm happy to do it in this PR if you can point me to a reference on how you are envision this.